Feature/gql variable

ashwin95r opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/171)
<!-- Reviewable:end -->

manishrjain commented:
I'm not convinced by the implementation. It touches too many functions. I think there should be a simpler way where you can do a post step of setting the values for the variables at the end, after generating GraphQuery; instead of doing it during the generation.

Think about how you might achieve that.

---

Reviewed 2 of 4 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[gql/parser.go, line 126 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPkN9HWq02h1qINbEaV-r1-126:-KPwn8_uu7qc6NLN7GrE:bi9xpvp) ([raw file](https://github.com/dgraph-io/dgraph/blob/0916e1b0bf6598d152bc67e21befc3e5ff35a80d/gql/parser.go#L126)):*
<details><summary><i>Previously, houndci-bot (Hound) wrote…</i></summary>
> exported type JsonQuery should have comment or be unexported<br>type JsonQuery should be JSONQuery
</details>
decode is a verb, not a noun. Such naming belongs to functions, not structs.

---

*[gql/parser.go, line 138 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwnZMbzNB7od-4T_0c:-KPwnZMbzNB7od-4T_0d:b-kfxr7x) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L138)):*
> ```Go
> func checkForVariableList(str string) (string, varMap, error) {
> 	var p decodeQuery
> 	tempMap := make(varMap)
> ```

vm := make(varMap)

---

*[gql/parser.go, line 141 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwnVfuZ4jqII5qovTL:-KPwnVfuZ4jqII5qovTM:b-jerag7) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L141)):*
> ```Go
> 	err := json.Unmarshal([]byte(str), &p)
> 	if err != nil {
> 		return str, tempMap, nil // It does not obey GraphiQL format but valid
> ```

return str, nil, nil?

---

*[gql/parser.go, line 152 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwnjIE6D-68XSA5q2x:-KPwnjIE6D-68XSA5q2y:bdoea9m) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L152)):*
> ```Go
> }
> 
> func checkValidityOfVariables(mp varMap) error {
> ```

checkValidityofVariables is redundant. Think about how it would be called. Something like: `checkValidity(vm)` is expressive enough.

---

*[gql/parser.go, line 157 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwo5NFSnSA6mWJne6t:-KPwo5NFSnSA6mWJne6u:bgp1ect) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L157)):*
> ```Go
> 
> 		// Ensure value is not nil if the variable is required
> 		if v.Type[len(v.Type)-1] == '!' {
> ```

Did you trim to ensure that you don't end up picking a space here?

---

*[gql/parser.go, line 161 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoDAUCgA-uz0KwcFm:-KPwoDAUCgA-uz0KwcFn:bfff25h) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L161)):*
> ```Go
> 				return fmt.Errorf("Variable %v should be initialised", k)
> 			}
> 			typ = v.Type[:len(v.Type)-1]
> ```

Trim here.

---

*[gql/parser.go, line 169 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoIVxeLd7ftZRAaGA:-KPwoIVxeLd7ftZRAaGB:b-b9q50t) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L169)):*
> ```Go
> 			case "int":
> 				{
> 					_, err := strconv.ParseInt(v.Value, 0, 64)
> ```

if _, err := statement; err != nil { .. }

---

*[gql/parser.go, line 176 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoLK7GIaI2NWg7WZ8:-KPwoLK7GIaI2NWg7WZ9:b-yq3bu8) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L176)):*
> ```Go
> 			case "float":
> 				{
> 					_, err := strconv.ParseFloat(v.Value, 64)
> ```

and here.

---

*[gql/parser.go, line 181 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoSHokf65ajEgtB5z:-KPwoSHokf65ajEgtB6-:b-yq3bu8) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L181)):*
> ```Go
> 					}
> 				}
> 			case "bool":
> ```

and here.

---

*[gql/parser.go, line 199 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPworTyMfz54vi0IHpl:-KPworTz1A5EEuW88FFW:b-f03js2) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L199)):*
> ```Go
> func Parse(input string) (gq *GraphQuery, mu *Mutation, rerr error) {
> 	l := &lex.Lexer{}
> 	query, vMap, err := checkForVariableList(input)
> ```

This is not checking, it's more like parsing. `parseQueryWithVariables(..)`

Also, I think you can just use varmap. And the variables could be called vmap. The camel case​ here looks a bit strange to my eyes, but if you want to stick to camel case, that's ok as well.

---

*[gql/parser.go, line 232 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwpMFhuF4bko4CprfY:-KPwpMFhuF4bko4CprfZ:b-d5b1p6) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L232)):*
> ```Go
> 		case itemLeftCurl:
> 			if gq == nil {
> 				if err = checkValidityOfVariables(vMap); err != nil {
> ```

if err = checkValidity(vmap); err != nil

---

*[gql/parser.go, line 245 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwplR5emg5wN71CziP:-KPwplR5emg5wN71CziQ:bz2xbn1) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L245)):*
> ```Go
> 			}
> 
> 		case itemLeftRound:
> ```

Can you directly have a left bracket here? Should it be preceded by query or something?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KPwrwjXCv52AV9YpO0Q:b-6vy9um)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
In fact, first, offset and after could themselves be converted to a variable map in GraphQuery. And fit within this generic implementation of variables. That way, we won't be doing special cases for them, while also simplifying the logic here.

---

Review status: all files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KPwspocqe5a1b4KCq-5:bw2k5s1)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Yeah, it could be moved post generation of GraphQuery, though it would lead to an extra parse of the query tree. Regarding putting the first, offset, after in a map, should it be a map of string to string, where types conversion is done later? And should we have a map in subgraph nodes as well?

---

Review status: 2 of 4 files reviewed at latest revision, 19 unresolved discussions.

---

*[gql/parser.go, line 126 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPkN9HWq02h1qINbEaV-r1-126:-KQ-Mmxpl3Mv6Gxzkt99:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/0916e1b0bf6598d152bc67e21befc3e5ff35a80d/gql/parser.go#L126)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> decode is a verb, not a noun. Such naming belongs to functions, not structs.
</details>
Done.

---

*[gql/parser.go, line 138 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwnZMbzNB7od-4T_0c:-KQ-DqpedAQIVGodqGsn:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L138)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> vm := make(varMap)
</details>
Done.

---

*[gql/parser.go, line 141 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwnVfuZ4jqII5qovTL:-KQ-NTqaYcWcMaw85mJZ:bqcxqf2) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L141)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> return str, nil, nil?
</details>
Returning an empty map would help avoid checking if it is nil at some other places.

---

*[gql/parser.go, line 152 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwnjIE6D-68XSA5q2x:-KQ-DyFEnoueIlSCXsj1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L152)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> checkValidityofVariables is redundant. Think about how it would be called. Something like: `checkValidity(vm)` is expressive enough.
</details>
Done.

---

*[gql/parser.go, line 157 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwo5NFSnSA6mWJne6t:-KQ-OPGYydQ-htqxugFG:b-6zqg58) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L157)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Did you trim to ensure that you don't end up picking a space here?
</details>
The lexer shouldn't be setting a space at the end as we return the item on encountering a space.

---

*[gql/parser.go, line 161 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoDAUCgA-uz0KwcFm:-KQ-PPbaqxFkk9WEcmR0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L161)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Trim here.
</details>
Done.

---

*[gql/parser.go, line 169 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoIVxeLd7ftZRAaGA:-KQ-E1MU7G9ScR0dHtFj:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L169)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> if _, err := statement; err != nil { .. }
</details>
Done.

---

*[gql/parser.go, line 176 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoLK7GIaI2NWg7WZ8:-KQ-E2HASIrVEQpgNfGz:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L176)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> and here.
</details>
Done.

---

*[gql/parser.go, line 181 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoSHokf65ajEgtB5z:-KQ-E2rywOqOeimZGsvM:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L181)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> and here.
</details>
Done.

---

*[gql/parser.go, line 199 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPworTyMfz54vi0IHpl:-KQ-EIk6iepxFfBN9_lC:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L199)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> This is not checking, it's more like parsing. `parseQueryWithVariables(..)`
> 
> Also, I think you can just use varmap. And the variables could be called vmap. The camel case​ here looks a bit strange to my eyes, but if you want to stick to camel case, that's ok as well.
</details>
Done.

---

*[gql/parser.go, line 232 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwpMFhuF4bko4CprfY:-KQ-D3P1YWXZONIqPa2x:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L232)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> if err = checkValidity(vmap); err != nil
</details>
Done.

---

*[gql/parser.go, line 245 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwplR5emg5wN71CziP:-KQ-cKXVVseN6LqAD_Ns:boifitg) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L245)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Can you directly have a left bracket here? Should it be preceded by query or something?
</details>
Actually it should be of the form `query something(...) { ... }`. Have done some modification in the way we handle queries. Let me know how it looks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KQ-ULnU2H1PUTQ0wnIQ:b9ni836)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Is an extra iteration of the generated query tree a concern? It would significantly improve code readability.

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[gql/parser.go, line 161 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoDAUCgA-uz0KwcFm:-KQ60xtC2t4L4MsAFi-w:bjr8z73) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L161)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary>
> Done.
</details>
That's not what I meant. Trim for `!` would still not trim for spaces.

---

*[gql/parser.go, line 245 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwplR5emg5wN71CziP:-KQ61Mvgyv1eakSDsMWF:bgxcx4j) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L245)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary>
> Actually it should be of the form `query something(...) { ... }`. Have done some modification in the way we handle queries. Let me know how it looks.
</details>
The word `query` is optional in GraphQL. I hope you didn't make it mandatory?

---

*[gql/parser.go, line 137 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQ60WlbE_c4q_nX-ALK:-KQ60WlbE_c4q_nX-ALL:ba86tq7) ([raw file](https://github.com/dgraph-io/dgraph/blob/fce53e3b5bb9b79388e07ca2d5886229e653c7f6/gql/parser.go#L137)):*
> ```Go
> 
> func parseQueryWithVariables(str string) (string, varMap, error) {
> 	var p query
> ```

var q query. Variable names tend to be initials or abbreviations of the structs.

---

*[gql/parser.go, line 152 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQ60pDsE9oOibhrj3DS:-KQ60pDsE9oOibhrj3DT:b-f17u3r) ([raw file](https://github.com/dgraph-io/dgraph/blob/fce53e3b5bb9b79388e07ca2d5886229e653c7f6/gql/parser.go#L152)):*
> ```Go
> }
> 
> func checkValidity(mp varMap) error {
> ```

vm varMap

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KQ61d9xBYuvub3oycIf:bsm2s41)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Modified the argumets to map.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[gql/parser.go, line 161 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoDAUCgA-uz0KwcFm:-KQ70KoywvcuGPRxdymb:bjzd21h) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L161)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> That's not what I meant. Trim for `!` would still not trim for spaces.
</details>
The lexer shouldn't include the spaces in these tokens. Though adding a trim above for safety.

---

*[gql/parser.go, line 245 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwplR5emg5wN71CziP:-KQ6z02gsJ5P1cgWO8uv:be28wi3) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L245)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> The word `query` is optional in GraphQL. I hope you didn't make it mandatory?
</details>
No, it still is optional.

---

*[gql/parser.go, line 137 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQ60WlbE_c4q_nX-ALK:-KQ6yCqqxuC-yW1NSJRe:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fce53e3b5bb9b79388e07ca2d5886229e653c7f6/gql/parser.go#L137)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> var q query. Variable names tend to be initials or abbreviations of the structs.
</details>
Done.

---

*[gql/parser.go, line 152 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQ60pDsE9oOibhrj3DS:-KQ6yDTq-WrGHklixckG:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/fce53e3b5bb9b79388e07ca2d5886229e653c7f6/gql/parser.go#L152)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> vm varMap
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KQ71CIRcbkcb8bFneTt:b8ul23y)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
This is shaping out to be a very nice change. A few more comments, and this should be ready to check in.

---

Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[gql/parser.go, line 161 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoDAUCgA-uz0KwcFm:-KQAlr7J_vHNDYKjJItW:bebxd89) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L161)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary>
> The lexer shouldn't include the spaces in these tokens. Though adding a trim above for safety.
</details>
If you don't expect it, then it's okay. You can remove the above line. But, then don't use the Trim function just to remove the single `!`. I think your original impl was better: typ = typ[0:len-1]

---

*[gql/parser.go, line 190 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAm_GUbHBGGBLgv_kl:-KQAm_GVrX6OTieArS9t:bipildz) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/gql/parser.go#L190)):*
> ```Go
> }
> 
> func substituteVariables(gq *GraphQuery, vmap varMap) error {
> ```

There is no error here. So, remove the return.

---

*[gql/parser.go, line 558 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAnQ99WUW3VNviCqYR:-KQAnQ9A9bIA8t_nhz02:b9haj4e) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/gql/parser.go#L558)):*
> ```Go
> 
> 		} else if item.Typ == itemName {
> 			child := new(GraphQuery)
> ```

@jchiu0 did a new thing recently where he started setting fields during the creation of variables. Like:

child := &GraphQuery{
  Args: make(..),
  Attr: item.Val,
}

I think it makes for a better reading code.

---

*[query/query.go, line 459 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAqMF_7TrFysME5_H6:-KQAqMF_7TrFysME5_H7:bj7gd0p) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/query/query.go#L459)):*
> ```Go
> 		}
> 		dst.Attr = gchild.Attr
> 		if gchild.Args["offset"] != "" {
> ```

Think you're looking for presence. Because you already check for empty fields.

if v, ok := gchild.Args["offset"]; ok {
  offset, err := strconv
  dst.Offset = int(offset)
}

---

*[query/query.go, line 466 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAqbNvvtnRzD5c7T7r:-KQAqbNw4XNzbwHs9b87:b-1kzc9c) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/query/query.go#L466)):*
> ```Go
> 			dst.Offset = int(offset)
> 		}
> 		if gchild.Args["after"] != "" {
> ```

And here.

---

*[query/query.go, line 473 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAqcUiJrAGZgVj-2xU:-KQAqcUiJrAGZgVj-2xV:b-1kzc9c) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/query/query.go#L473)):*
> ```Go
> 			dst.AfterUid = uint64(after)
> 		}
> 		if gchild.Args["first"] != "" {
> ```

And here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KQAsdP-cIYOTFEtjYpb:b-32bmmz)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 2 of 5 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[gql/parser.go, line 161 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KPwoDAUCgA-uz0KwcFm:-KQJQm7NUpseCbC9Ygq3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5c674d73861ee795c3fa438ee75283524df4e18/gql/parser.go#L161)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> If you don't expect it, then it's okay. You can remove the above line. But, then don't use the Trim function just to remove the single `!`. I think your original impl was better: typ = typ[0:len-1]
</details>
Done.

---

*[gql/parser.go, line 190 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAm_GUbHBGGBLgv_kl:-KQJNMrGmj3Wml84NY1t:bsf2s6n) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/gql/parser.go#L190)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> There is no error here. So, remove the return.
</details>
Actually, We should have a check to ensure that the variable exists in map. Added it now.

---

*[gql/parser.go, line 558 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAnQ99WUW3VNviCqYR:-KQJO3XvjHRr9q1dQ2nS:b-294uzj) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/gql/parser.go#L558)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> @jchiu0 did a new thing recently where he started setting fields during the creation of variables. Like:
> 
> child := &GraphQuery{
>   Args: make(..),
>   Attr: item.Val,
> }
> 
> I think it makes for a better reading code.
</details>
Ture. Done.

---

*[query/query.go, line 459 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAqMF_7TrFysME5_H6:-KQJPtigFrWZfTXVVXEp:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/query/query.go#L459)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Think you're looking for presence. Because you already check for empty fields.
> 
> if v, ok := gchild.Args["offset"]; ok {
>   offset, err := strconv
>   dst.Offset = int(offset)
> }
</details>
Done.

---

*[query/query.go, line 466 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAqbNvvtnRzD5c7T7r:-KQJPuOKLaLrPhOOO2zv:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/query/query.go#L466)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> And here.
</details>
Done.

---

*[query/query.go, line 473 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQAqcUiJrAGZgVj-2xU:-KQJPv3Mg_EqP7-q_hta:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/b5278140283e23888ada2f42afc2334755f39f0d/query/query.go#L473)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> And here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[gql/parser.go, line 182 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJfNMP5TAF2AavZ_Ua:-KQJfNMP5TAF2AavZ_Ub:b-igq147) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L182)):*
> ```Go
> 						return fmt.Errorf("Expected a bool but got %v", v.Value)
> 					}
> 				}
> ```

Also, handle default: Possibly returning an error that we haven't implemented the type yet?

---

*[gql/parser.go, line 191 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJa2462GF4IojfyfEf:-KQJa2462GF4IojfyfEg:b9qylrl) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L191)):*
> ```Go
> 
> func substituteVariables(gq *GraphQuery, vmap varMap) error {
> 	for k, v := range gq.Args {
> ```

What if `v` is empty? Is it being checked somewhere? If so, add a comment here. Otherwise, check here.

---

*[gql/parser.go, line 202 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJdw3lAhIpEgO98o0F:-KQJdw3ntTqm96LAsoOC:blz2o0x) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L202)):*
> ```Go
> 
> 	for _, child := range gq.Children {
> 		substituteVariables(child, vmap)
> ```

Your current impl is lossy about errors. Check for errors here and return that error.

---

*[gql/parser.go, line 258 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJe68wzl32o6IQ8mNt:-KQJe68wzl32o6IQ8mNu:bzbmz7l) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L258)):*
> ```Go
> 			return nil, nil, err
> 		}
> 		substituteVariables(gq, vmap)
> ```

Repeat after me: I will always check for errors. Every time I don't check for errors, Chennai gets flooded.

---

*[gql/parser.go, line 264 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJfCtW2tspYVMB3AxV:-KQJfCtW2tspYVMB3AxW:b-5wyz02) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L264)):*
> ```Go
> }
> 
> func getVariablesAndQuery(l *lex.Lexer, vmap varMap) (gq *GraphQuery, rerr error) {
> ```

Roughly explain how this function operates. Which tokens is it looking for , etc. in a comment.

---

*[gql/parser.go, line 299 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJf9Cqz037DXDE0N94:-KQJf9Cqz037DXDE0N95:b-he9kgh) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L299)):*
> ```Go
> }
> 
> func getQuery(l *lex.Lexer) (gq *GraphQuery, rerr error) {
> ```

What does this function do? Add a comment.

---

*[gql/parser.go, line 411 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJesaNE4bbrfMAnBm-:-KQJesaNE4bbrfMAnBm0:bvv7tsm) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L411)):*
> ```Go
> }
> 
> func parseVariables(l *lex.Lexer, vmap varMap) error {
> ```

Can you add comments within this func to explain the flow of code?

---

*[gql/parser.go, line 453 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJeomB7sJ83e6NRFUL:-KQJeomB7sJ83e6NRFUM:b-puf53q) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L453)):*
> ```Go
> 			}
> 
> 			if vmap[varName].Value == "" {
> ```

Looking for presence or empty string?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 3 of 5 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[gql/parser.go, line 182 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJfNMP5TAF2AavZ_Ua:-KQJsvNH1Ife6397SYzh:b426qny) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L182)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Also, handle default: Possibly returning an error that we haven't implemented the type yet?
</details>
Yeah, we can throw an error saying the type is not supported. As of now, we support `int, float, bool, string`

---

*[gql/parser.go, line 191 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJa2462GF4IojfyfEf:-KQJgwVQZ0vY0DDSteVg:bq4hjvq) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L191)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> What if `v` is empty? Is it being checked somewhere? If so, add a comment here. Otherwise, check here.
</details>
Yes. It is being checked in parseVariables function.

---

*[gql/parser.go, line 202 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJdw3lAhIpEgO98o0F:-KQJhQLLGHRAibeURVd1:b-71hn04) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L202)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Your current impl is lossy about errors. Check for errors here and return that error.
</details>
Ah sorry, my bad. Done.

---

*[gql/parser.go, line 258 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJe68wzl32o6IQ8mNt:-KQJhTaAc1fGz-_EceSm:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L258)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Repeat after me: I will always check for errors. Every time I don't check for errors, Chennai gets flooded.
</details>
Done.

---

*[gql/parser.go, line 264 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJfCtW2tspYVMB3AxV:-KQJioO_pNRg9YpTS3mT:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L264)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Roughly explain how this function operates. Which tokens is it looking for , etc. in a comment.
</details>
Done.

---

*[gql/parser.go, line 299 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJf9Cqz037DXDE0N94:-KQJkiw1wGc7v4eZYNth:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L299)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> What does this function do? Add a comment.
</details>
Done.

---

*[gql/parser.go, line 411 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJesaNE4bbrfMAnBm-:-KQJmMVPlNnuDFyEhIEW:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L411)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Can you add comments within this func to explain the flow of code?
</details>
Done.

---

*[gql/parser.go, line 453 \[r8\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQJeomB7sJ83e6NRFUL:-KQJmGKGtvIrimwQ1AZk:b-30ilvb) ([raw file](https://github.com/dgraph-io/dgraph/blob/3bdef0864cd6efeeb0f21c3c9ac888a903c04388/gql/parser.go#L453)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Looking for presence or empty string?
</details>
Checking if it is empty. Actually, for string both would be equivalent as nil for string is "".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Looks good! Nice change. One small comment about a spelling mistake. Also, noticed some comments which don't finish with a dot. Just fix these before submitting.

Add a task for yourself to update the wiki about these queries.

---

Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[gql/parser.go, line 453 \[r9\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-KQKGEFPP6kOPKfWv_P6:-KQKGEFPP6kOPKfWv_P7:b-9xw87w) ([raw file](https://github.com/dgraph-io/dgraph/blob/7e5c43ccc1d6b142818b0cd47812e6aa03f625e6/gql/parser.go#L453)):*
> ```Go
> 		}
> 
> 		// Insert the variable into the map. The variable might already be defiend
> ```

defined

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KQKHbVP0DqsLVese4l4:br95ywa)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Sure, Fixed. Added a task for wiki: https://app.asana.com/0/110655897041979/173240972017079

---

Review status: 4 of 5 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/171#-:-KQKTP1dOW9doJbJwmJx:bh9xi0o)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Merged to master with commit f526cf499e4f1caef955d7b839501dbbaaa0a8d8.
